### PR TITLE
eq-868 Additional Analytics Events

### PR DESCRIPTION
**Pull request by hamishtaplin**
### What is the context of this PR?

[Trello card](https://trello.com/c/qAG9cWeP/868-additional-analytics-events)

screenshot 2017-03-10 12 36 34

Adds additional event tracking for: 

- Error screens
- Events when the user uses any Navigation links
- Events when the timeout dialog appears
- Events when the session expires
- Events when the user clicks 'Save and complete later'

I've refactored `analytics.js` so that it uses a more declarative style for event tracking. Instead of selecting the elements and event/data to track in the JS, it now look for DOM elements and tracks accordingly. Implemented so far:

- Clicks (when a user clicks an element)
- Errors (when an error is on the page)
- Visible (when an element is or becomes visible)

The system works via a jinja helper macro `track` which takes the event type, category, action and label as parameters and maps them to `data` attributes in the DOM.
 
### How to review 

Check each type of event is being tracked:
- errors
- guidance
- previous link
- 'save and complete later'
- timeout popup
- timeout expiry
- section navigation
- help and support

### CODE REVIEW

File: app/templates/partials/answer-guidance.html [hunk 1]
```diff
@@ -15,11 +15,13 @@
...
 {% set guidance_body = {
   "id": "guidance-" ~ answer_guidance.id ~ "-body",
+  "id": "guidance-" ~ answer_guidance.id ~ "-body",
   "aria-hidden": "true"
 } %}
 
 <div class="guidance js-details" {{guidance|xmlattr}}>
-  <a class="guidance__link js-details-trigger js-details-label mars" {{guidance_link|xmlattr}}>{{_("Show further guidance")}}</a>
+  <a class="guidance__link js-details-trigger js-details-label mars"
+    {{guidance_link|xmlattr}} {{helpers.track('click', 'Navigation', 'Guidance', answer_guidance.id)}}>{{_("Show further guidance")}}</a>
```
Comment (collisdigital) [inline app/templates/partials/answer-guidance.html]: Maybe the category for this should be something else; 'Help' or 'Content' or something?

File: app/templates/partials/navigation.html [hunk 2]
```diff
@@ -6,7 +7,8 @@
       {% if nav_item.current_location %}
         <div>{{nav_item.link_name}}</div>
       {% else %}
-        <a class="nav__link" href="{{nav_item.link_url}}">{{nav_item.link_name}}</a>
+        <a class="nav__link" href="{{nav_item.link_url}}"
+          {{helpers.track('click', 'Navigation', 'Section link', nav_item.link_url)}}>{{nav_item.link_name}}</a>
```
Comment (collisdigital) [inline app/templates/partials/navigation.html]: Did you decide not to make the action's verbs in the end? e.g. `Section Link Clicked` ?

File: app/themes/census/templates/partials/header.html [hunk 3]
```diff
@@ -13,7 +14,11 @@
     </div>
   </div>
   <div class="header__container container">
-    <div class="header__logo"><img src="{{ url_for('static', filename='img/theme/census/census-logo.svg') }}" alt="2017 Census" /></div><button class="header__btn btn btn--light btn--border js-help-btn js-help-and-support u-hidden">
+    <div class="header__logo">
+      <img src="{{ url_for('static', filename='img/theme/census/census-logo.svg') }}" alt="2017 Census" />
+    </div>
+    <button class="header__btn btn btn--light btn--border js-help-btn u-hidden"
+      {{helpers.track('click', 'Navigation', 'Help and support')}}>
```
Comment (collisdigital) [inline app/themes/census/templates/partials/header.html]: Not sure about the category of this being Navigation?

File: gulp/scripts.js [hunk 4]
```diff
@@ -31,6 +31,9 @@ const b = browserify({
 .on('update', () => {
   cache = bundle()
 })
+.on('error', () => {
+  gutil.log('sdfdsffsfs')
```
Comment (collisdigital) [inline gulp/scripts.js]: this meant to be here?

File: app/templates/partials/questions/errors.html [hunk 5]
```diff
@@ -1,8 +1,9 @@
+{% import 'macros/helpers.html' as helpers %}
 <div class="panel panel--simple panel--error" data-qa="error">
   <div class="panel__header">
     {% if form.question_errors[question.id] is not none %}
     <ul class="list list--bare list--errors">
-      <li class="list__item venus" {{helpers.errorTracking(question.id, error)}}>{{ form.question_errors[question.id] }}</li>
+      <li class="list__item venus" {{helpers.track('error', 'Error', error, question.id)}}>{{ form.question_errors[question.id] }}</li>
```
Comment (collisdigital) [inline app/templates/partials/questions/errors.html:6]: Should this be Error category with 'Error displayed' as the action and the label as the error?

File: app/templates/partials/answer-guidance.html (hunk 1 above)
Comment (hamishtaplin) [reply app/templates/partials/answer-guidance.html]: Yeah I did wonder if 'Help' would be a category

File: app/templates/partials/navigation.html (hunk 2 above)
Comment (hamishtaplin) [reply app/templates/partials/navigation.html]: It felt a bit redundant but I can add it back in if necessary

File: app/themes/census/templates/partials/header.html (hunk 3 above)
Comment (hamishtaplin) [reply app/themes/census/templates/partials/header.html]: 'Help' category I think

File: gulp/scripts.js (hunk 4 above)
Comment (hamishtaplin) [reply gulp/scripts.js]: yeah thats a new JS framework called 'sdfdsffsfs.js'

Comment (collisdigital) [reply gulp/scripts.js]: It wouldn't surprise me 🙄

File: app/templates/partials/questions/errors.html (hunk 5 above)
Comment (hamishtaplin) [reply app/templates/partials/questions/errors.html:6]: That would be more consistent but don't we want to have the error message (action) as well as the question.id (label)?

**Review by bennpearson — APPROVED** [review]

File: app/templates/partials/answer-guidance.html [hunk 6]
```diff
@@ -15,11 +15,13 @@
 
 {% set guidance_body = {
   "id": "guidance-" ~ answer_guidance.id ~ "-body",
+  "id": "guidance-" ~ answer_guidance.id ~ "-body",
```
Comment (ajmaddaford) [inline app/templates/partials/answer-guidance.html]: Looks like a duplicated line here?

File: app/templates/partials/answer-guidance.html [hunk 7]
```diff
@@ -15,11 +15,13 @@
...
   "aria-hidden": "true"
 } %}
 
 <div class="guidance js-details" {{guidance|xmlattr}}>
-  <a class="guidance__link js-details-trigger js-details-label mars" {{guidance_link|xmlattr}}>{{_("Show further guidance")}}</a>
+  <a class="guidance__link js-details-trigger js-details-label mars"
+    {{guidance_link|xmlattr}} {{helpers.track('click', 'Help', 'Guidance click', answer_guidance.id)}}>{{_("Show further guidance")}}</a>
```
Comment (ajmaddaford) [inline app/templates/partials/answer-guidance.html:23]: Is it ok/good practice to have a new line within the `<a>` tag?

File: app/templates/partials/navigation.html [hunk 8]
```diff
@@ -6,7 +7,8 @@
       {% if nav_item.current_location %}
         <div>{{nav_item.link_name}}</div>
       {% else %}
-        <a class="nav__link" href="{{nav_item.link_url}}">{{nav_item.link_name}}</a>
+        <a class="nav__link" href="{{nav_item.link_url}}"
+          {{helpers.track('click', 'Navigation', 'Section link click', nav_item.link_url)}}>{{nav_item.link_name}}</a>
```
Comment (ajmaddaford) [inline app/templates/partials/navigation.html:11]: Is it ok/good practice to have a new line within the `<a>` tag?

File: app/templates/partials/previous-link.html [hunk 9]
```diff
@@ -0,0 +1,6 @@
+{% import 'macros/helpers.html' as helpers %}
+{% if previous_location %}
+  <a class="page__previous page__previous--{{position}}"
+  {{helpers.track('click', 'Navigation', 'Previous link click', position)}}
+  href="{{previous_location}}" id="{{position}}-previous">Previous</a>
```
Comment (ajmaddaford) [inline app/templates/partials/previous-link.html:5]: Is it ok/good practice to have a new line within the `<a>` tag?

**Review by ajmaddaford — CHANGES_REQUESTED** [review]

**Comment by collisdigital** [comment]
Reviewed actual events going into GA with @hamishtaplin and @ajmaddaford; looks good. Code review comments, squash and re-base needed.

File: app/templates/partials/answer-guidance.html (hunk 7 above)
Comment (hamishtaplin) [reply app/templates/partials/answer-guidance.html:23]: It doesn't matter as far as I know as browsers ignore them

**Review by ajmaddaford — APPROVED** [review]

**Pull request merged**